### DISCUSSION TITLE
i18n and SEO - some considerations

### DISCUSSION THREAD

**Discussion by milovangudelj** [vercel/next.js#51712] [App Router]
The currently recommended approach to internationalization leaves some gaps and undesirable side effects.

## Crawlers
The biggest issue is that search engine crawlers fail inspection of any URL they attempt to visit. From what I can see in the logs they repeatedly hit an internal server error caused by `@formatjs`'s locale matcher (`Incorrect locale information provided`). As a result Google is not able to index your pages and you won't see page previews on deployments.

As an example see the following images:

![image](https://github.com/vercel/next.js/assets/49202538/88c5ccdd-7db1-4bd4-96b0-e5f2e39b7ce9)
![image](https://github.com/vercel/next.js/assets/49202538/4d53c1b8-9a68-41e5-a9bd-1f360dd8dcda)
![image](https://github.com/vercel/next.js/assets/49202538/1fbf72bc-2a92-412c-a1b5-d360bc01db22)

I don't really know what headers are present on the crawler's request and what causes the error. It works perfectly both in development and production with any of the most popular browsers.

You can also test this by running the following command in your terminal 
```fish
curl --request GET --url https://www.milovangudelj.com
```

***EDIT:***

Nevermind! :sweat_smile:
A fix for this is to tell `Negotiator` what locales are available when you call `Negotiator.languages()` like so:
```ts
let languages = new Negotiator({ headers: negotiatorHeaders }).languages(
   locales
);
```

## Redirects & Rewrites

This is up to personal preference really, but I'll put it here anyway.
Prefixing all your pages with a locale leaves you without a base URL and you end up with this scenario:
```node
GET "https://www.milovangudelj.com" --[redirects to]--> "https://www.milovangudelj.com/en/"  --[rewrites to]--> "https://www.milovangudelj.com"
```
***Note***: The example doesn't tell you how to do this. It just leaves the default locale visible in the address bar.

I haven't had the time to try and really work on this yet, but for me an ideal setup would not redirect if you visit the canonical URL and match the default locale. This would need a totally different approach than generating all possible page variants with `generateStaticParams` including the default locale.

## Sitemap & SEO
It would be nice if you provided an example sitemap that takes into account canonical and alternate links. Maybe this is being planned as a future docs improvement once the full `sitemap.ts` API is available, but as of today if you want this you're basically on your own.

The way I did this was by leveraging middleware. On each request if the pathname matches `/sitemap.xml` I return the result of a custom `generateSiteMap()` function that takes care of everything, from alternate links to dynamic routes like for example `/work/[slug]`.

## Conclusion
Maybe I'm doing it the wrong way. It could be. If so please share what you did, what issues you encountered and how you solved them.

If you have any suggestions and ideas please contribute to the discussion by commenting below and you might inadvertently help someone or just give them a nudge in the right direction.

**Comment by sayhicoelho** [comment]
Try this https://github.com/svobik7/next-roots
